### PR TITLE
No error/help display for JSON output

### LINKED ISSUE #6098: `odo <invalid-command> -o json` returns both human readable and json output
/kind bug

## What versions of software are you using?

**Operating System:** Fedora 32

**Output of `odo version`:** odo v3.0.0-rc1 (9d72e39bf8)

## How did you run odo exactly?
Running odo with a parent command and json returns both json and human readable output.
1. `odo describe -o json` or any invalid command

## Actual behavior
```sh
$ odo describe -o json
Error: Subcommand not found, use one of the available commands: binding, component
Usage:
  odo describe [flags]
  odo describe [command]

Available Commands:
  binding     Describe bindings
  component   Describe a component

Flags:
  -h, --help   Help for describe

Additional Flags:
      --kubeconfig string    Paths to a kubeconfig. Only required if out-of-cluster.
  -o, --o string             Specify output format, supported format: json (default "")
  -v, --v Level              Number for the log level verbosity. Level varies from 0 to 9 (default 0).
      --vmodule moduleSpec   Comma-separated list of pattern=N settings for file-filtered logging

Use "odo describe [command] --help" for more information about a command. 

{
        "message": "Subcommand not found, use one of the available commands: binding, component"
}
```

## Expected behavior

```sh
$ odo describe -o json
{
        "message": "Subcommand not found, use one of the available commands: binding, component"
}
```
Should only return output that's asked for.

### PR DESCRIPTION
**Pull request by feloy**
**What type of PR is this:**

/kind bug

**What does this PR do / why we need it:**

**Which issue(s) this PR fixes:**

Fixes #6098

**PR acceptance criteria:**

- [x] Unit test 

- [ ] Integration test 

- [x] Documentation 

**How to test changes / Special notes to the reviewer:**

### CODE REVIEW

**Comment by feloy** [comment]
The first commit fixes this kind of command:
```
odo qwe -o json
```

But not:

```
odo describe -o json
```

**Review by valaparthvi — COMMENTED** [review]
/approve

The code lgtm, do you think we should check an integration test just in case there's a regression in the future?

**Comment by feloy** [comment]
> /approve
> 
> The code lgtm, but do you think we should check an integration test just in case there's a regression in the future?

Sure, I'll add some integration tests for this

**Comment by feloy** [comment]
This PR does not fix the problem for such commands (commands for which we have optional sub-commands):

```
$ odo list unknown -o json
```

**Comment by valaparthvi** [comment]
> This PR does not fix the problem for such commands (commands for which we have optional sub-commands):
> 
> ```
> $ odo list unknown -o json
> ```

IIUC, this error originates from the `NoArgs` that we use for `odo list`.
```go
// NewCmdList implements the list odo command
func NewCmdList(name, fullName string) *cobra.Command {
	o := NewListOptions()

	var listCmd = &cobra.Command{
		...
		...
		Args: func(cmd *cobra.Command, args []string) error {
			if log.IsJSON() {
				cmd.SilenceUsage = true
				cmd.SilenceErrors = true
			}
			return cobra.NoArgs(cmd, args)
		},
		...
		...
	}
...
...
```
Modifying the `Args` to silence errors for JSON works.

**Comment by feloy** [comment]
> > This PR does not fix the problem for such commands (commands for which we have optional sub-commands):
> > ```
> > $ odo list unknown -o json
> > ```
> 
> IIUC, this error originates from the `NoArgs` that we use for `odo list`.
> 
> ```go
> // NewCmdList implements the list odo command
> func NewCmdList(name, fullName string) *cobra.Command {
> 	o := NewListOptions()
> 
> 	var listCmd = &cobra.Command{
> 		...
> 		...
> 		Args: func(cmd *cobra.Command, args []string) error {
> 			if log.IsJSON() {
> 				cmd.SilenceUsage = true
> 				cmd.SilenceErrors = true
> 			}
> 			return cobra.NoArgs(cmd, args)
> 		},
> 		...
> 		...
> 	}
> ...
> ...
> ```
> 
> Modifying the `Args` to silence errors for JSON works. But we'll need to modify all commands using `NoArgs`, I think we can define our own `NoArgs` function and use that.

Great, thanks for the hint, I was not able to find at which place to set these variables to true.

I've replaced all NoArgs with a specific function

File: tests/integration/generic_test.go [hunk 1]
```diff
@@ -52,6 +52,22 @@ var _ = Describe("odo generic", func() {
...
+
+	})
+
+	It("returns JSON error when using an invalid sub-command with JSON output", func() {
+		res := helper.Cmd("odo", "describe", "unknown-sub-command", "-o", "json").ShouldFail()
+		stdout, stderr := res.Out(), res.Err()
+		Expect(stdout).To(BeEmpty())
+		Expect(helper.IsJSON(stderr)).To(BeTrue())
+
+	})
```
Comment (valaparthvi) [inline tests/integration/generic_test.go:69]: Perhaps you can combine these error messages in a single It spec to reduce setup overhead?
There's one more case of `odo describe -o json` where `odo describe` cannot function as a standalone command.

**Review by valaparthvi — APPROVED** [review]

**Comment by feloy** [comment]
/override windows-integration-test/Windows-test

Unrelated failing tests with insufficient memory:
```
3 Insufficient memory
```

**Comment by feloy** [comment]
/override ci/prow/v4.10-integration-e2e
Tests pass on IBM Cloud

**Pull request merged**